pisi/fetcher.py

```diff
@@ -36,58 +38,53 @@ class FetchError(pisi.Error):
     pass
 
 
-class UIHandler:
-    def __init__(self, progress):
-        self.filename        = None
-        self.url             = None
-        self.basename        = None
-        self.downloaded_size = 0
-        self.percent         = None
-        self.rate            = 0.0
-        self.size            = 0
-        self.eta             = '--:--:--'
-        self.symbol          = '--/-'
-        self.last_updated    = 0
-        self.exist_size      = 0
-
-    def start(self, archive, url, basename, size, text):
+class FetchHandler:
+    def __init__(self, url, archive, bandwidth_limit):
+        self.url                = url
+        self.percent            = None
+        self.rate               = 0.0
+        self.size               = 0
+        self.eta                = '--:--:--'
+        self.symbol             = '--/-'
+        self.last_updated       = 0
+        self.filename           = url.filename()
+        self.total_size         = 0.0
+        self.exist_size         = 0
+        self.bandwidth_limit    = bandwidth_limit
         if os.path.exists(archive):
             self.exist_size = os.path.getsize(archive)
-        self.filename   = util.remove_suffix(ctx.const.partial_suffix, basename)
-        self.url        = url
-        self.basename   = basename
-        self.total_size = size or 0
-        self.text       = text
 
-        self.now    = lambda: time.time()
+        self.now = lambda: time.time()
         self.t_diff = lambda: self.now() - self.s_time
 
         self.s_time = self.now()
 
-    def update(self, size):
-
-        if self.size == size:
-            return
-
-        self.size = size
+    def update(self, blocknum, bs, size):
+        self.total_size = size + self.exist_size
+        self.size = blocknum * bs + self.exist_size
         if self.total_size:
-            self.percent = (size * 100.0) / self.total_size
+            self.percent = self.size * 100 / self.total_size
```

Comment:
Should pull in https://github.com/pisilinux/pisi/commit/c71327fbc0c05ad0df7719e1d2c1d844782fa224#diff-2a3b1b35244f1433b428f9205b2251d4f64b733abb7808254a3d4080df7b123e as well